library/alloc/src/collections/btree/node.rs

```diff
@@ -473,72 +499,87 @@ impl<'a, K, V, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
         NodeRef { height: self.height, node: self.node, _marker: PhantomData }
     }
 
+    /// Exposes the leaf portion of any leaf or internal node for writing.
+    ///
+    /// Returns a raw ptr to avoid invalidating other references to this node.
+    fn as_leaf_mut_ptr(&mut self) -> *mut LeafNode<K, V> {
+        self.node.as_ptr()
+    }
+
     /// Exposes the leaf portion of any leaf or internal node for writing.
     ///
     /// We don't need to return a raw ptr because we have unique access to the entire node.
     fn as_leaf_mut(&mut self) -> &'a mut LeafNode<K, V> {
         unsafe { &mut (*self.node.as_ptr()) }
     }
 
-    /// Borrows a mutable reference to one of the keys stored in the node.
+    /// Borrows a mutable reference to one of the key storage areas in the node.
     ///
     /// # Safety
     /// The node has more than `idx` initialized elements.
-    unsafe fn key_mut_at(&mut self, idx: usize) -> &mut K {
-        unsafe { self.reborrow_mut().into_key_mut_at(idx) }
+    unsafe fn key_area_mut_at(&mut self, idx: usize) -> &mut MaybeUninit<K> {
+        unsafe { self.reborrow_mut().into_key_area_mut_at(idx) }
     }
 
-    /// Borrows a mutable reference to one of the values stored in the node.
+    /// Borrows a mutable reference to one of the value storage areas in the node.
     ///
     /// # Safety
     /// The node has more than `idx` initialized elements.
-    unsafe fn val_mut_at(&mut self, idx: usize) -> &mut V {
-        unsafe { self.reborrow_mut().into_val_mut_at(idx) }
+    unsafe fn val_area_mut_at(&mut self, idx: usize) -> &mut MaybeUninit<V> {
+        unsafe { self.reborrow_mut().into_val_area_mut_at(idx) }
     }
 
-    fn keys_mut(&mut self) -> &mut [K]
+    /// Borrows a reference to a stretch of key storage area in the node.
+    ///
+    /// # Safety
+    /// - The node has more than `idx` initialized elements.
+    fn keys_mut(&mut self, start: usize) -> &mut [MaybeUninit<K>]
```

Comment:
This function is safe but has a safety comment?

Also, the function is actually not safe, as I can do awful things like
```rust
foo.keys_mut()[0] = MaybeUninit::uninit();
```

Reply:
I fixed that already, wait a second...

Reply:
Is there any caller that does *not* use 0 for `start`? I don't entirely understand why you added that parameter.

Reply:
> Also, the function is actually not safe, as I can do awful things like

But that's basically what `slice_remove` does, and what MaybeUninit is for, so is that really unsafe? It's only unsafe when you later `assume_init` it, right?

Reply:
The concern about the return type remains -- functions returning `&mut MaybeUninit` are rarely safe because they allow the caller to *de-initialize* the given memory.

Reply:
> But that's basically what slice_remove does, and what MaybeUninit is for, so is that really unsafe? It's only unsafe when you later assume_init it, right?

Doesn't de-initializing violate the invariant of the node? The idea of a safe function is that safe clients can do what they want with it without violating any safety invariants.

IOW, it *will* be assume_init'ed later by safe code doing a map lookup.

Reply:
Yes it violates the invariant, but internal code has to do that to change pretty much anything.

I understand your concern, but that has applied to other internal functions like `as_leaf` for a long time? As it happens, I tagged that unsafe now, but not because of the mess you can cause using its return value, but because it might stump on other references.

Reply:
I agree that for internal functions it can be less clear-cut who is responsible for upholding which invariant.

> but because it might stump on other references.

it doesn't do anything odd with lifetimes though? So that seems more like something the caller needs to be aware of regardless to me.^^

But in the end it is somewhat arbitrary, so as long as it is properly documented I can live with any of these choices here.

Reply:
>  I tagged that unsafe now

No I didn't. I forgot to tag `as_leaf_mut`, and if I try that, the avalanche means you're tempted to make all bodies unsafe.

Reply:
So I made the `as_X` methods safe again, and made the ones returning `MaybeUninit` references and slices unsafe. It doesn't make much sense, since you can access the `MaybeUninit` array directly and safely. But you need an unsafe block to do use them anyway.

Maybe the fields shouldn't be open to the entire node code, and the methods returning `MaybeUninit` stuff should be members of `LeafNode`/`InternalNode` instead of `NodeRef`. I fenced off the `BoxedNode` field just to see what the reviewer(s) will say.

Reply:
One class of methods that has been very clear to me from the start, is those on a plain `self` like `first_edge`, `left_kv`: they simply return the same ownership, borrowship, or whatever you call it.

Next to the easy ones are those on a plain `self` called `into_X` (where I think the unwritten rule is that X is a part of the original, not a neighbour). Some of them return an `'a`  lifetime, which stems from the `NodeRef`'s `marker::Immut<'a>`/`ValMut<'a>`/`Mut<'a>`. It didn't understand what that implied for a long time, but they didn't look daunting either, thanks to the plain `self`.

I now understand why there is both `into_key_at` and `key_at`: because the former profits from the lifetime incorporated into `marker::Immut<'a>`/`ValMut<'a>`/`Mut<'a>`, and the latter is defined and used for all 3 of those, and the only way to do that is to define it over some `BorrowType` without lifetime. Therefore you can write the latter using the former, but not the other way around. `into_key_at` is easy enough and more sophisticated.

Therefore I turned all the new/changed function into the `into_X` class, and I like it.

Reply:
If there's a good place to document that structure with a comment in the code, that seems like it could be helpful. :)